ktor-client/ktor-client-cio/common/src/io/ktor/client/engine/cio/utils.kt

```diff
@@ -102,38 +104,52 @@ internal suspend fun readResponse(
     val rawResponse = parseResponse(input)
         ?: throw EOFException("Failed to parse HTTP response: unexpected EOF")
 
-    val status = HttpStatusCode(rawResponse.status, rawResponse.statusText.toString())
-    val contentLength = rawResponse.headers[HttpHeaders.ContentLength]?.toString()?.toLong() ?: -1L
-    val transferEncoding = rawResponse.headers[HttpHeaders.TransferEncoding]
-    val connectionType = ConnectionOptions.parse(rawResponse.headers[HttpHeaders.Connection])
+    rawResponse.use {
+        val status = HttpStatusCode(rawResponse.status, rawResponse.statusText.toString())
+        val contentLength = rawResponse.headers[HttpHeaders.ContentLength]?.toString()?.toLong() ?: -1L
+        val transferEncoding = rawResponse.headers[HttpHeaders.TransferEncoding]
+        val connectionType = ConnectionOptions.parse(rawResponse.headers[HttpHeaders.Connection])
 
-    val headers = buildHeaders {
-        appendAll(CIOHeaders(rawResponse.headers))
-        rawResponse.headers.release()
-    }
+        val rawHeaders = rawResponse.headers
+        val headers = HeadersImpl(rawHeaders.toMap())
+        val version = HttpProtocolVersion.parse(rawResponse.version)
 
-    val version = HttpProtocolVersion.parse(rawResponse.version)
+        if (status == HttpStatusCode.SwitchingProtocols) {
+            return startWebSocketSession(status, requestTime, headers, version, callContext, input, output)
+        }
 
-    if (status == HttpStatusCode.SwitchingProtocols) {
-        return startWebSocketSession(status, requestTime, headers, version, callContext, input, output)
-    }
+        val body = when {
+            request.method == HttpMethod.Head ||
+                status in listOf(HttpStatusCode.NotModified, HttpStatusCode.NoContent) ||
+                status.isInformational() -> {
+                ByteReadChannel.Empty
+            }
+            else -> {
+                val httpBodyParser = GlobalScope.writer(Dispatchers.Unconfined, autoFlush = true) {
+                    parseHttpBody(contentLength, transferEncoding, connectionType, input, channel)
```

Comment:
A little bit "slippery" here since we pass pooled values here to a launched coroutine and the only hope is that it is unconfined. How was release done in before this change? Was the response object released at the context completion? Wasn't it the issue before this change too since the coroutine is running outside of the hierarchy?
❓  Also, we have the same issue with potential spaces in the header so should we fix it everywhere or we give up and don't care.